tests/aws-cpp-sdk-core-tests/smithy/client/SmithyClientTest.cpp

```diff
@@ -19,16 +27,162 @@ class SmithyClientTest : public Aws::Testing::AwsCppSdkGTestSuite {
+
+
+class TestClient : public MySmithyClient
+{
+    public:
+    TestClient(const MySmithyClientConfig& clientConfig,
+            const Aws::String& serviceName,
+            const std::shared_ptr<Aws::Http::HttpClient>& httpClient,
+            const std::shared_ptr<Aws::Client::AWSErrorMarshaller>& errorMarshaller,
+            const std::shared_ptr<MyTestEndpointProvider> endpointProvider,
+            const std::shared_ptr<MyServiceAuthSchemeResolver>& authSchemeResolver,
+            const Aws::UnorderedMap<Aws::String, SigVariant>& authSchemesMap):
+            MySmithyClient(
+                clientConfig,
+                serviceName,
+                httpClient,
+                errorMarshaller,
+                endpointProvider,
+                authSchemeResolver,
+                authSchemesMap)
+    {}
+
+
+    SelectAuthSchemeOptionOutcome SelectAuthSchemeOption(const smithy::client::AwsSmithyClientAsyncRequestContext& ctx) const
+    {
+        return MySmithyClient::SelectAuthSchemeOption(ctx);
+    }
+
+    SigningOutcome SignRequest(std::shared_ptr<HttpRequest> httpRequest, const smithy::AuthSchemeOption& targetAuthSchemeOption) const
+    {
+        return MySmithyClient::SignRequest(httpRequest, targetAuthSchemeOption);
+    }
+
+
+};
+
+
+
+TEST_F(SmithyClientTest, testSigV4) {
+
+    const char ALLOCATION_TAG[] = "SmithyClientTest";
+
+    Aws::Client::ClientConfiguration clientConfig;
+    std::shared_ptr<Aws::Http::HttpClient> httpClient = Aws::Http::CreateHttpClient(clientConfig);
+    std::shared_ptr<Aws::Client::AWSErrorMarshaller> errorMarshaller = Aws::MakeShared<Aws::Client::XmlErrorMarshaller>(ALLOCATION_TAG);
+    std::shared_ptr<MyTestEndpointProvider> endPointProvider;
+
+    std::shared_ptr<MyServiceAuthSchemeResolver> authSchemeResolver = Aws::MakeShared<smithy::SigV4AuthSchemeResolver<> >(ALLOCATION_TAG);
+
+
+    Aws::UnorderedMap<Aws::String, SigVariant> authSchemesMap;
+
+    smithy::SigV4AuthScheme::SigV4AuthSchemeParameters params;
+    params.serviceName = "MyService";
+    params.region = "us-west2";
+    params.operation = "TestOperation";
+
+    
+    Aws::String key{"aws.auth#sigv4"};
+
+    SigVariant val{smithy::SigV4AuthScheme(params)};
+    
+    authSchemesMap.emplace(key, val);
+
+    std::shared_ptr<TestClient> ptr = Aws::MakeShared<TestClient>(
+        ALLOCATION_TAG,
+        clientConfig,
+        "MyService",
+        httpClient,
+        errorMarshaller,
+        endPointProvider,
+        authSchemeResolver,
+        authSchemesMap
+        );
+    smithy::client::AwsSmithyClientAsyncRequestContext ctx;
+    ctx.m_pRequest = nullptr;
+
+    auto res = ptr->SelectAuthSchemeOption(ctx);
+
+    EXPECT_EQ(res.IsSuccess(), true);
+    std::cout<<"selected scheme id="<<res.GetResult().schemeId<<std::endl;
+    EXPECT_EQ(res.GetResult().schemeId, key);
+
+    //for s3 client as an example
+    Aws::String uri{"https://objecttest-cb93079d-24a0-4862-8ef2-88dae12cd549.s3.amazonaws.com"};
```

Comment:
lets use a test string instead of one generated by integration tests

Reply:
I can, but my ultimate goal is to test the entire auth workflow where i figured it needs a real valid endpoint for the integration test

Reply:
yes but _this is a real bucket_ if you want to integration test a existing bucket, you have to create/destroy it as part of the test

Reply:
ack

Reply:
This is just a mock string now, i used it to preserve the uri format

Reply:
still valid in this revision, lets make this a fake bucket name in the URI

Reply:
ack, let me stub out keeping the format